kube-core/src/version.rs

```diff
@@ -198,33 +270,55 @@ mod tests {
     }
 
     #[test]
-    fn test_version_ord() {
+    fn test_version_latest_stable_ord() {
         // sorting makes sense from a "greater than" semantic perspective:
-        assert!(Version::Stable(2) > Version::Stable(1));
-        assert!(Version::Stable(1) > Version::Beta(1, None));
-        assert!(Version::Stable(1) > Version::Beta(2, None));
-        assert!(Version::Stable(2) > Version::Alpha(1, Some(2)));
-        assert!(Version::Stable(1) > Version::Alpha(2, Some(2)));
-        assert!(Version::Beta(1, None) > Version::Nonconformant("ver3".into()));
-
-        assert!(Version::Stable(2) > Version::Stable(1));
-        assert!(Version::Stable(1) > Version::Beta(2, None));
-        assert!(Version::Stable(1) > Version::Beta(2, Some(2)));
-        assert!(Version::Stable(1) > Version::Alpha(2, None));
-        assert!(Version::Stable(1) > Version::Alpha(2, Some(3)));
-        assert!(Version::Stable(1) > Version::Nonconformant("foo".to_string()));
-        assert!(Version::Beta(1, Some(1)) > Version::Beta(1, None));
-        assert!(Version::Beta(1, Some(2)) > Version::Beta(1, Some(1)));
-        assert!(Version::Beta(1, None) > Version::Alpha(1, None));
-        assert!(Version::Beta(1, None) > Version::Alpha(1, Some(3)));
-        assert!(Version::Beta(1, None) > Version::Nonconformant("foo".to_string()));
-        assert!(Version::Beta(1, Some(2)) > Version::Nonconformant("foo".to_string()));
-        assert!(Version::Alpha(1, Some(1)) > Version::Alpha(1, None));
-        assert!(Version::Alpha(1, Some(2)) > Version::Alpha(1, Some(1)));
-        assert!(Version::Alpha(1, None) > Version::Nonconformant("foo".to_string()));
-        assert!(Version::Alpha(1, Some(2)) > Version::Nonconformant("foo".to_string()));
-        assert!(Version::Nonconformant("bar".to_string()) > Version::Nonconformant("foo".to_string()));
-        assert!(Version::Nonconformant("foo1".to_string()) > Version::Nonconformant("foo10".to_string()));
+        assert!(Version::Stable(2).latest_stable() > Version::Stable(1).latest_stable());
+        assert!(Version::Stable(1).latest_stable() > Version::Beta(1, None).latest_stable());
```

Comment:
hmm. i'm not a big fan of this api to be honest. the whole point of this `Version` struct is to sort it by the version priority, not to make version priority a second rate citizen of it.

if we want to have standard semver ordering (and i don't know why we would), then we can implement something else for it eventually.

Reply:
hm.. maybe the `latest()` sort key by itself can make sense as an alternate way to sort, if you think there is a use case for sorting by latest?

Reply:
> hmm. i'm not a big fan of this api to be honest. the whole point of this Version struct is to sort it by the version priority, not to make version priority a second rate citizen of it.

To me, this is not about making priority sorting a second-class citizen, but to acknowledge that there is no single universal sorting method, and that the user should be aware of whatever choice they're making.

Would you be more comfortable with renaming `latest` to `latest_prerelease`, or something else like that?

> the whole point of this Version struct is to sort it by the version priority

I don't think that's obvious at all from the name `kube::core::Version`.

Reply:
> Would you be more comfortable with renaming `latest` to `latest_prerelease`, or something else like that?
> 
> > the whole point of this Version struct is to sort it by the version priority
> 
> I don't think that's obvious at all from the name `kube::core::Version`.

That is a fair point, and I am a little less hesitant to go for two sort orders if the names can help clarify this. I.e. something like:

- `Version::priority -> impl Ord`
- `Version::latest` -> impl Ord`

> To me, this is not about making priority sorting a second-class citizen, but to acknowledge that there is no single universal sorting method, and that the user should be aware of whatever choice they're making.

In general I would agree with you, but we have a struct here with 4 different version types, with only 3 of which we can reasonably invent a `latest` order for. As soon as someone puts a nonconformant order in a list with a beta release (say), we are not going to have expected behaviour no matter what we do. This is why I think we have a canonical order. The `Version` struct forces us to follow it.

Reply:
> That is a fair point, (and am open to suggestions for making it more obvious), but am a little less hesitant to go for two sort orders if the names can help clarify this. I.e. something like:

Those sound good to me. d2c36c75b7e0b506924fd555614cf01948b98e41

> As soon as someone puts a nonconformant order in a list with a beta release (say), we are not going to have expected behaviour no matter what we do. This is why I think we have a canonical order.

Agreed, but we have that problem for both orderings, and regardless of whether we consider one to be canonical.

I think that splitting out the non-conformant case is the only way to get somewhere close to reasonable in that regard.

Reply:
Looks better. :+1:

> Agreed, but we have that problem for both orderings, and regardless of whether we consider one to be canonical.

It's not a problem for the priority sort because it's part of the spec, but it is problematic for the latest sort.

Maybe it's not such a big deal to have the latest sort do the same as priority sort in this case (i.e. sort nonconformant after versioned ones - they are unlikely to coexist in a list unless people produce both non-conformant and conformant versions - which is a user error).

Given that the standard api complexity has grown now from:

1. `.sort_by_cached_key(|gvd| Version::parse(gvd.version.as_str()))` (master)
2. `.sort_by_cached_key(|gvd| Reverse(Version::parse(gvd.version.as_str())))` (mine)
3. `.sort_by_cached_key(|gvd| Reverse(Version::parse(gvd.version.as_str()).priority()))` (this currently)
4. `.sort_by_cached_key(|gvd| Reverse(Version::parse(gvd.version.as_str()).map(Version::priority))` (further proposed splitting)

I am keen to not have it go much further. I feel like step 4 may be a little bit too far here as well. But going to sleep on it.

Reply:
(I think hesitance to a Result type here is probably because `Nonconformant` is not an error with the priority type, but it kind of is an error with the latest type - and having different apis for the two sorts would probably be undesirable?)